File: tracks/java/src/Lasagna.java
/**
 * The Lasagna class is used to calculate various times associated with the
 * preparation and cooking of lasagna.
 */
public class Lasagna {
    /**
     * Default constructor for Lasagna.
     * This constructor doesn't perform any special operations and is provided
     * to create instances of this class if needed.
     */
    public Lasagna() {

    }

    /**
     * Returns the standard expected oven time in minutes for cooking lasagna.
     *
     * @return the expected oven time in minutes
     */
    public int expectedMinutesInOven() {
        return 40;
    }

    /**
     * Calculates the remaining minutes in the oven based on the actual minutes
     * already spent in the oven.
     *
     * @param actualMinuteInOven the number of minutes the lasagna has already been
     *                           in the oven
     * @return the number of remaining minutes required in the oven
     */
    public int remainingMinutesInOven(int actualMinuteInOven) {
        return expectedMinutesInOven() - actualMinuteInOven;
    }

    /**
     * Calculates the preparation time in minutes based on the number of layers.
     * Assumes a fixed time to prepare each layer.
     *
     * @param layerCount the number of lasagna layers
     * @return the preparation time in minutes
     */
    public int preparationTimeInMinutes(int layerCount) {
        return layerCount * 2;
    }

    /**
     * Calculates the total time in minutes spent on preparing and cooking the
     * lasagna.
     *
     * @param layerCount         the number of lasagna layers
     * @param actualMinuteInOven the number of minutes the lasagna has already been
     *                           in the oven
     * @return the total time in minutes spent on the lasagna
     */
    public int totalTimeInMinutes(int layerCount, int actualMinuteInOven) {
        return preparationTimeInMinutes(layerCount) + actualMinuteInOven;
    }
}

File: tracks/python/lasagna.py
"""Functions used in preparing Guido's gorgeous lasagna.

Learn about Guido, the creator of the Python language:
https://en.wikipedia.org/wiki/Guido_van_Rossum

This is a module docstring, used to describe the functionality
of a module and its functions and/or classes.
"""


EXPECTED_BAKE_TIME = 40
PREPARATION_TIME = 2


def bake_time_remaining(elapsed_bake_time):
    """Calculate the bake time remaining.

    :param elapsed_bake_time: int - baking time already elapsed.
    :return: int - remaining bake time (in minutes) derived from 'EXPECTED_BAKE_TIME'.

    Function that takes the actual minutes the lasagna has been in the oven as
    an argument and returns how many minutes the lasagna still needs to bake
    based on the `EXPECTED_BAKE_TIME`.
    """

    return EXPECTED_BAKE_TIME - elapsed_bake_time


def preparation_time_in_minutes(number_of_layers):
    """Calculate the preparation time in minutes.

    :param number_of_layers: int - number of layers to be prepared.
    :return: int - time to bake the layers (in minutes) derived from 'PREPARATION_TIME'.

    Function that takes the number of layers as an argument and returns 
    how many minutes all the layers need to be prepared 
    based on the `PREPARATION_TIME`.
    """

    return PREPARATION_TIME * number_of_layers


def elapsed_time_in_minutes(number_of_layers, elapsed_bake_time):
    """
    :param number_of_layers: int - The number of layers to prepare.
    :elapsed_bake_time: int - The amount of time the baking has already taken in minutes.
    :return: int - The total elapsed time in minutes.

    Calculate the total elapsed time in minutes.

    This function sums the total preparation time (determined by the number of layers) and 
    the elapsed bake time to calculate the total elapsed time since the start of the baking 
    process. 
    """
    return preparation_time_in_minutes(number_of_layers) + elapsed_bake_time
